File: vigenere/sic/vigenere.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include <ctype.h>

char* extend(char*,int);
int main(int argc, char** argv) {
	char* string = argv[1];
	char* key = argv[2];
	
	//printf("%s\n", string);
	//printf("%s\n", key);
	
	float slen = strlen(string);
	float klen = strlen(key);
	
	char* ref;
	if (slen > klen) {
		int duplicate = floor(ceil(slen/klen));
		ref = extend(key, duplicate);
	} else if (slen < klen) {
		char* destination = malloc(strlen(string)+1);
		ref = strncat(destination,key,strlen(string));
	} else {
		ref = key;
	}
	//printf("%s\n", ref);
	
	int i;
	char *rtn = malloc(strlen(string)+1);
	char nchar;
	
	for(i=0; i<strlen(string);i++) {
		nchar = string[i];
		if (isalpha(nchar)) {
			char schar = string[i]-'a';
			char kchar = ref[i]-'a';
			nchar = (schar+kchar)%26 + 'a';
		}
		rtn[i] = nchar;
	}
	
	printf("%s\n", rtn);
}

char* extend(char* key, int repeat) {
	char* temp = key;
	key = malloc(strlen(key)*repeat+1);
	strcpy(key, temp);
	while (repeat > 1) {
		strcat(key, temp);
		repeat--;
	}
	return key;
}

File: vigenere/sic/vigenere.py
#!/usr/bin/python

from sys import argv, exit
import math

def main():
	string = argv[1].lower()
	key = argv[2].lower()
	
	slen = len(string)
	klen = len(key)
	
	if slen > klen:
		newkey = key*(int(math.ceil(float(slen)/float(klen))))
		if len(newkey) > slen:
			key = newkey[0:slen]
		else:
			key = newkey
	elif slen < klen:
		key = key[0:slen]

	s = ""
	for i in range(0,slen):
		schar = string[i]
		kchar = key[i]
		nchar = " "
		if schar.isalpha():
			sord = ord(schar)-ord('a')
			kord = ord(kchar)-ord('a')
			nord = (sord+kord)%26+ord('a')
			nchar = chr(nord)
		else:
			nchar = schar
		s += nchar
	
	print s
		

if __name__ == '__main__':
	main()
